src/algebra/module/torsion.lean

```diff
@@ -0,0 +1,207 @@
+
+instance i : has_scalar S (torsion' R M S) := ⟨λ s x, ⟨s • x, begin
+  obtain ⟨x, a, h⟩ := x, use a, dsimp, rw [smul_comm, h, smul_zero] end⟩⟩
+instance : distrib_mul_action S (torsion' R M S) := subtype.coe_injective.distrib_mul_action
+  ((torsion' R M S).subtype).to_add_monoid_hom (λ (c : S) x, rfl)
+instance : smul_comm_class S R (torsion' R M S) := ⟨λ s a x, begin ext, exact smul_comm _ _ _ end⟩
+
+/-- The `S`-torsion submodule of the `S`-torsion submodule (viewed as a module) is the full
+`S`-torsion module. -/
+lemma torsion'_torsion'_eq_top : torsion' R (torsion' R M S) S = ⊤ :=
+eq_top_iff.mpr (λ ⟨x, ⟨a, h⟩⟩ _, ⟨a, subtype.ext h⟩)
+/-- The torsion submodule of the torsion submodule (viewed as a module) is the full
+torsion module. -/
+lemma torsion_torsion_eq_top : torsion R (torsion R M) = ⊤ := torsion'_torsion'_eq_top R⁰
+
+end torsion'
+
+section torsion
+variables [comm_semiring R] [add_comm_monoid M] [no_zero_divisors R] [nontrivial R] [module R M]
+
+example (a : R⁰) (x : M) : a • x = (a : R) • x := submonoid.smul_def a x
+
+lemma coe_torsion_eq_annihilator_ne_bot :
+  (torsion R M : set M) = { x : M | (R ∙ x).annihilator ≠ ⊥ } :=
+begin
+  ext x, simp_rw [submodule.ne_bot_iff, mem_annihilator, mem_span_singleton],
+  exact ⟨λ ⟨a, hax⟩, ⟨a, λ _ ⟨b, hb⟩, by rw [← hb, smul_comm, ← submonoid.smul_def, hax, smul_zero],
+    non_zero_divisors.coe_ne_zero _⟩,
+    λ ⟨a, hax, ha⟩, ⟨⟨_, mem_non_zero_divisors_of_ne_zero ha⟩, hax x ⟨1, one_smul _ _⟩⟩⟩
+end
+
+/-- A module over a domain has `no_zero_smul_divisors` iff its torsion submodule is trivial. -/
+lemma no_zero_smul_divisors_iff_torsion_eq_bot :
+  no_zero_smul_divisors R M ↔ torsion R M = ⊥ :=
+begin
+  split; intro h,
+  { haveI : no_zero_smul_divisors R M := h,
+    rw eq_bot_iff, rintro x ⟨a, hax⟩,
+    change (a : R) • x = 0 at hax,
+    cases eq_zero_or_eq_zero_of_smul_eq_zero hax with h0 h0,
+    { exfalso, exact non_zero_divisors.coe_ne_zero a h0 }, { exact h0 } },
+  { exact { eq_zero_or_eq_zero_of_smul_eq_zero := λ a x hax, begin
+      by_cases ha : a = 0,
+      { left, exact ha },
+      { right, rw [← mem_bot _, ← h],
+        exact ⟨⟨a, mem_non_zero_divisors_of_ne_zero ha⟩, hax⟩ }
+    end } }
+end
+end torsion
+
+namespace quotient_torsion
+variables [comm_ring R] [add_comm_group M] [module R M]
+
+/-- Quotienting by the torsion submodule gives a torsion-free module. -/
+lemma torsion_eq_bot : torsion R (M ⧸ torsion R M) = ⊥ :=
+eq_bot_iff.mpr $ λ z, quotient.induction_on' z $ λ x ⟨a, hax⟩,
+begin
+  rw [quotient.mk'_eq_mk, ← quotient.mk_smul, quotient.mk_eq_zero] at hax,
+  rw [mem_bot, quotient.mk'_eq_mk, quotient.mk_eq_zero],
+  cases hax with b h,
+  exact ⟨b * a, (mul_smul _ _ _).trans h⟩
+end
+
+
+instance no_zero_smul_divisors [is_domain R] : no_zero_smul_divisors R (M ⧸ torsion R M) :=
+no_zero_smul_divisors_iff_torsion_eq_bot.mpr torsion_eq_bot
+
+end quotient_torsion
+end
+end submodule
+
+section ℕ_torsion
```

Comment:
In general, we should have a predicate `module.is_torsion`. But that can wait for a follow-up PR.